python/pyarrow/tests/test_pandas.py

```diff
@@ -2882,6 +2882,55 @@ def test_variable_dictionary_with_pandas():
         a.to_pandas()
 
 
+# ----------------------------------------------------------------------
+# Array protocol in pandas conversions tests
+
+
+@pytest.mark.skipif(LooseVersion(pd.__version__) < '0.24.0',
+                    reason='IntegerArray only introduced in 0.24')
+def test_array_protocol():
+
+    def __arrow_array__(self, type=None):
+        return pa.array(self._data, mask=self._mask, type=type)
+
+    df = pd.DataFrame({'a': pd.Series([1, 2, None], dtype='Int64')})
+
+    # with latest pandas/arrow, trying to convert nullable integer errors
+    with pytest.raises(TypeError):
+        pa.table(df)
+
+    # patch IntegerArray with the protocol
+    pd.arrays.IntegerArray.__arrow_array__ = __arrow_array__
+
+    # default conversion
+    result = pa.table(df)
+    expected = pa.array([1, 2, None], pa.int64())
+    assert result[0].chunk(0).equals(expected)
+
+    # with specifying schema
+    schema = pa.schema([('a', pa.float64())])
+    result = pa.table(df, schema=schema)
+    expected2 = pa.array([1, 2, None], pa.float64())
+    assert result[0].chunk(0).equals(expected2)
+
+    # pass Series to pa.array
+    result = pa.array(df['a'])
+    assert result.equals(expected)
+    result = pa.array(df['a'], type=pa.float64())
+    assert result.equals(expected2)
+
+    # pass actual ExtensionArray to pa.array
+    result = pa.array(df['a'].values)
+    assert result.equals(expected)
+    result = pa.array(df['a'].values, type=pa.float64())
+    assert result.equals(expected2)
+
+    try:
+        del pd.arrays.IntegerArray.__arrow_array__
```

Comment:
This cleanup clause should be in a `try..finally` block IMHO.

Reply:
Sorry, I am not sure I follow. What should be in the `finally` block? 
You mean putting the whole testing code above in the "try" block and the cleanup in the finally?

In that case, I could maybe use a pytest fixture with yield instead?

Reply:
> You mean putting the whole testing code above in the "try" block and the cleanup in the finally?

Yes.

> In that case, I could maybe use a pytest fixture with yield instead?

I kindof hate pytest fixtures :-S

Reply:
OK, updated with a try/finally

> I kindof hate pytest fixtures :-S

I also dislike over-use of them (certainly if they are magically coming from other files), but for within a single file, they can have their use :-) But since I am only using the pattern a single time, not worth to create a function for it this time.